### PR TITLE
Feature/dependency pinning and update automation

### PR DESCRIPTION
**Pull request by antoooks**
- Enable `gorepomod` to work outside $GOSRC, by adding `--local`
- Enable auto fetch update on `gorepomod list` and `gorepomod pin`
- Enable gorepomod to work on fork repository
- Ensure backward compatibility and manual execution still works
- Add test for local flag
- Enable release for `patch`
- Add backward compatibility for `unpin`

`Note:` implementation done based on assumption that the release process will produce `release-{module}-vX.Y.Z` release branch naming. And this branch is created manually when releasing which will become trigger to the CI steps

Addresses #5404

### CODE REVIEW

**Comment by koba1t** [comment]
@antoooks 
Looks like CI failed. Could you fix the Lint error?

File: cmd/gorepomod/internal/git/runner.go [hunk 1]
```diff
@@ -165,8 +167,20 @@ func (gr *Runner) LoadLocalTags() (result misc.VersionMap, err error) {
 
 func (gr *Runner) LoadRemoteTags(
 	remote misc.TrackedRepo) (result misc.VersionMap, err error) {
-	gr.comment("loading remote tags")
 	var out string
+
+	// Update latest tags from upstream
+	gr.comment("updating tags from upstream")
+	_, err = gr.run(noHarmDone, "fetch", "-t", string(remoteUpstream), "master")
```
Comment (koba1t) [inline cmd/gorepomod/internal/git/runner.go:174]: I think the `master` branch name maybe rename, I suggest using const variable to define the upstream default branch name.
https://github.com/kubernetes-sigs/kustomize/issues/3954

File: cmd/gorepomod/internal/repo/dotgitdata.go [hunk 2]
```diff
@@ -136,3 +169,33 @@ func (dg *DotGitData) checkModules(modules []*protoModule) error {
...
+	}
+
+	out, err := exec.Command("git", "config", "--get", "remote.origin.url").Output()
+	if err != nil {
+		fmt.Errorf("failed extracting git information: %q", err.Error())
+	}
+
+	var originUrlString string = string(out)
+	var localPrefix string = utils.ParseGitRepositoryPath(originUrlString)
+	return localPrefix
```
Comment (koba1t) [inline cmd/gorepomod/internal/repo/dotgitdata.go:200]: Do you have any special reason to use a few temporary variables?
I think we can write one line from the above 3 lines, and I suggest using any comment line if you need to explain what is happening.

File: cmd/gorepomod/internal/utils/utils.go [hunk 3]
```diff
@@ -45,3 +47,38 @@ func SliceContains(slice []string, target string) bool {
...
+	// for ssh protocol git@github.com:path/repo.git
+	switch protocol {
+	case "git":
+		repoPath = strings.Join(d[2:len(d)-1], "/") + "/" + d[3][:len(d[3])-4]
+		// https protocol https://github.com/path/repo.git
+	case "https":
+		repoPath = strings.Join(d[2:], "/")
+		repoPath = repoPath[:len(repoPath)-4]
+		// unsupported format
+		// TODO: confirm if we should handle other formats not commonly supported by Github
```
Comment (koba1t) [inline cmd/gorepomod/internal/utils/utils.go:68]: Please write your GitHub id in the TODO comment.
https://github.com/kubernetes-sigs/kustomize/blob/0b4974eb1ce0316c7b1c2f82365c858b5c9b37d4/kyaml/openapi/openapi.go#L675

This ID is super useful when tracking TODO.

**Review by koba1t — CHANGES_REQUESTED** [review]
Thanks for your work!
I reviewed this PR. Please check my comments.

**Comment by koba1t** [comment]
This PR looks almost good!
I will check the behavior of the gorepomod command after a few small comments are fixed.

File: cmd/gorepomod/internal/repo/dotgitdata.go (hunk 2 above)
Comment (antoooks) [reply cmd/gorepomod/internal/repo/dotgitdata.go:200]: None in particular, I forgot to refactor it after debugging

**Comment by antoooks** [comment]
It's weird, I ran lint locally and it came out fine. But if it ran on CI it throws many check errors

Screenshot 2023-12-07 at 8 12 31 AM

@koba1t do you know how I can reproduce same linting rule as CI?

**Comment by koba1t** [comment]
Hi @antoooks 
I try to pin and unpin kyaml without `go/src` dir. So looks like the `go.mod` file was broken.
Could you investigate this?

```
kob@Yugos-MacBook-Air kustomize % git diff HEAD
diff --git a/api/go.mod b/api/go.mod
index c350266f7..66d98517b 100644
--- a/api/go.mod
+++ b/api/go.mod
@@ -36,4 +36,4 @@ require (
        gopkg.in/yaml.v3 v3.0.1 // indirect
 )

-replace sigs.k8s.io/kustomize/kyaml => ../kyaml
+replace kustomize/kyaml => ../kyaml
diff --git a/api/go.sum b/api/go.sum
index efb9f252a..30ee0ec8c 100644
```

**Comment by antoooks** [comment]
> Hi @antoooks I try to pin and unpin kyaml without `go/src` dir. So looks like the `go.mod` file was broken. Could you investigate this?
> 
> ```
> kob@Yugos-MacBook-Air kustomize % git diff HEAD
> diff --git a/api/go.mod b/api/go.mod
> index c350266f7..66d98517b 100644
> --- a/api/go.mod
> +++ b/api/go.mod
> @@ -36,4 +36,4 @@ require (
>         gopkg.in/yaml.v3 v3.0.1 // indirect
>  )
> 
> -replace sigs.k8s.io/kustomize/kyaml => ../kyaml
> +replace kustomize/kyaml => ../kyaml
> diff --git a/api/go.sum b/api/go.sum
> index efb9f252a..30ee0ec8c 100644
> ```

@koba1t Nice catch, I assumed we use separate release branch so I didn't work on unpin, but I think I should to support backward compatibility. Will fix it now, thanks Yugo-san

**Comment by koba1t** [comment]
> Nice catch, I assumed we use separate release branch so I didn't work on unpin, but I think I should to support backward compatibility.

Thanks, @antoooks!
Maybe this command is working on my computer. It looks good to me!
Let's try it in the next release.

/lgtm
/approve

**Review by koba1t — APPROVED** [review]

**Pull request merged**